api/v1alpha1/databasecluster_types.go

```diff
@@ -349,8 +349,27 @@ type DatabaseClusterSpec struct {
 	Monitoring *Monitoring `json:"monitoring,omitempty"`
 	// Sharding is the sharding configuration. PSMDB-only
 	Sharding *Sharding `json:"sharding,omitempty"`
+	// Affinity defines scheduling affinity per component.
+	Affinity []*DatabaseClusterAffinity `json:"affinity,omitempty"`
 }
 
+type DatabaseClusterAffinity struct {
+	// Component defines the component to which the affinity applies.
+	Component DatabaseClusterAffinityComponent `json:"type"`
+	// Affinity defines scheduling affinity.
+	Affinity corev1.Affinity `json:"affinity,omitempty"`
```

Comment:
This approach also works but I think it would be tidier if instead of having this `DatabaseClusterAffinity` struct with a component field we'd instead take this affinity field and added it to each component's struct (`Engine`, `Proxy`, `Sharding.ConfigServer`). 
Just like you can configure the resources for Engine and Proxy and the Resources field is part of those structs. I envisioned the Affinity to also be part of those structs. This is what I mean:
```
// Engine is the engine configuration.
type Engine struct {
	// Type is the engine type
	// +kubebuilder:validation:Enum:=pxc;postgresql;psmdb
	Type EngineType `json:"type"`
	// Version is the engine version
	Version string `json:"version,omitempty"`
	// Replicas is the number of engine replicas
	// +kubebuilder:validation:Minimum:=1
	Replicas int32 `json:"replicas,omitempty"`
	// Storage is the engine storage configuration
	Storage Storage `json:"storage"`
	// Resources are the resource limits for each engine replica.
	// If not set, resource limits are not imposed
	Resources Resources `json:"resources,omitempty"`
	// Config is the engine configuration
	Config string `json:"config,omitempty"`
	// UserSecretsName is the name of the secret containing the user secrets
	UserSecretsName string `json:"userSecretsName,omitempty"`
	// CRVersion is the desired version of the CR to use with the
	// underlying operator.
	// If unspecified, everest-operator will use the same version as the operator.
	//
	// NOTE: Updating this property post installation may lead to a restart of the cluster.
	// +optional
	CRVersion *string `json:"crVersion,omitempty"`
	// Affinity defines scheduling affinity.
	Affinity corev1.Affinity `json:"affinity,omitempty"`
}

// Proxy is the proxy configuration.
type Proxy struct {
	// Type is the proxy type
	// +kubebuilder:validation:Enum:=mongos;haproxy;proxysql;pgbouncer
	Type ProxyType `json:"type,omitempty"`
	// Replicas is the number of proxy replicas
	// +kubebuilder:validation:Minimum:=1
	Replicas *int32 `json:"replicas,omitempty"`
	// Config is the proxy configuration
	Config string `json:"config,omitempty"`
	// Expose is the proxy expose configuration
	Expose Expose `json:"expose,omitempty"`
	// Resources are the resource limits for each proxy replica.
	// If not set, resource limits are not imposed
	Resources Resources `json:"resources,omitempty"`
	// Affinity defines scheduling affinity.
	Affinity corev1.Affinity `json:"affinity,omitempty"`
}

// ConfigServer represents the sharding configuration server settings.
type ConfigServer struct {
	// Replicas is the amount of configServers
	// +kubebuilder:validation:Minimum:=1
	Replicas int32 `json:"replicas"`
	// Affinity defines scheduling affinity.
	Affinity corev1.Affinity `json:"affinity,omitempty"`
}
```

Reply:
I also thought about this approach at first but then figured there's more than 3 affinity configurations per engine.
For example MongoDB has 5.

I was thinking this may be a more flexible and easier to understand approach for users to managing affinity?
I'm happy to put it into the 3 structs though if that's preferred.

Reply:
I see your point, however with the current API we can't even enable the nonvoting/arbiter nodes so we don't really need to consider setting affinity for them. If we ever add support for configuring the nonvoting/arbiter nodes we'll need to expose it somehow which can then also have the affinity configuration there.

> easier to understand approach for users

I actually disagree here, I think that it's easier to understand an `Affinity` field embedded in `Engine`, `Proxy`, `Sharding.ConfigServer` than to see an `Affinity` array at the root of the spec where you then need to know how to name each component in order to fill in `Affinity.Component`.
Let's hear from @oksana-grishchenko and @mayankshah1607 maybe they feel differently than me :slightly_smiling_face: I'm open to changing my mind.

Reply:
I agree with Diogo, placing it into its 3 respective structs is also consistent with the current spec

Reply:
updated